runtime/sema/type_tags.go

```diff
@@ -0,0 +1,708 @@
+		}
+	case neverTypeMask:
+		return NeverType
+	case characterTypeMask:
+		return CharacterType
+	case boolTypeMask:
+		return BoolType
+	case voidTypeMask:
+		return VoidType
+	case addressTypeMask:
+		return &AddressType{}
+	case metaTypeMask:
+		return MetaType
+	case blockTypeMask:
+		return BlockType
+	case deployedContractMask:
+		return DeployedContractType
+	case pathTypeMask:
+		return PathType
+	case privatePathTypeMask:
+		return PrivatePathType
+	case publicPathTypeMask:
+		return PublicPathType
+	case storagePathTypeMask:
+		return StoragePathType
+	case capabilityPathTypeMask:
+		return CapabilityPathType
+	case anyStructTypeMask:
+		return AnyStructType
+	case anyResourceTypeMask:
+		return AnyResourceType
+	case anyTypeMask:
+		return AnyType
+	case noTypeMask:
+		return InvalidType
+
+	case compositeTypeMask:
+		// We reach here if all are composite types.
+		// Therefore, check for member types, and decide the
+		// common supertype based on the member types.
+		var prevType Type
+		for _, typ := range types {
+			// Ignore 'Never' type as it doesn't affect the supertype.
+			if typ == NeverType {
+				continue
+			}
+
+			if prevType == nil {
+				prevType = typ
+				continue
+			}
+
+			if !typ.Equal(prevType) {
+				return commonSuperTypeOfComposites(types)
+			}
+		}
+
+		return prevType
+
+	// All derived types goes here.
+	case constantSizedTypeMask,
+		variableSizedTypeMask,
+		dictionaryTypeMask,
+		referenceTypeMask,
+		genericTypeMask,
+		functionTypeMask,
+		interfaceTypeMask:
+
+		return getSuperTypeOfDerivedTypes(types)
+	default:
+		// not homogenous. Return nil and continue on advanced checks.
+		return nil
+	}
+}
+
+func findSuperTypeFromUpperMask(joinedTypeTag TypeTag, types []Type) Type {
+	switch joinedTypeTag.upperMask {
+
+	case invalidTypeMask:
+		return InvalidType
+
+	// All derived types goes here.
+	case capabilityTypeMask,
+		restrictedTypeMask,
+		transactionTypeMask:
+		return getSuperTypeOfDerivedTypes(types)
+	default:
+		return nil
+	}
+}
+
+func getSuperTypeOfDerivedTypes(types []Type) Type {
+	// We reach here if all types belongs to same kind.
+	// e.g: All are arrays, all are dictionaries, etc.
+	// Therefore, check for member types, and decide the
+	// common supertype based on the member types.
+	var prevType Type
+	for _, typ := range types {
+		// 'Never' type doesn't affect the supertype.
+		// Hence, ignore them
+		if typ == NeverType {
+			continue
+		}
+
+		if prevType == nil {
+			prevType = typ
+			continue
+		}
+
+		if !typ.Equal(prevType) {
+			return commonSuperTypeOfHeterogeneousTypes(types)
+		}
+	}
+
+	if prevType == nil {
+		return InvalidType
+	}
+
+	return prevType
+}
+
+func commonSuperTypeOfHeterogeneousTypes(types []Type) Type {
+	var hasStructs, hasResources bool
+	for _, typ := range types {
+		isResource := typ.IsResourceType()
+		hasResources = hasResources || isResource
+		hasStructs = hasStructs || !isResource
+
+		if hasResources && hasStructs {
+			return AnyType
```

Comment:
Yes, this is actually handled in the subsequent PR: https://github.com/onflow/cadence/pull/1027/files#diff-a28ba45130151972e99d735035c6afd277d716d92a12b6aa870238be0f0496e6R351-R360

Reply:
It also has some tests for similar cases:
https://github.com/onflow/cadence/blob/cde2d93c24342131d1c53b565f18dd016bafe482/runtime/tests/checker/type_inference_test.go#L902-L912

Reply:
Ah, I see, perfect. Given that this PR goes into `master`, could you please already include the check from the referenced PR into this PR? I worry we might forget and end up having the in-between state in a release

Reply:
I think it should be OK, since this PR only introduces the utility method `LeastCommonSuperType` but never uses it anywhere in checker/interpreter. PR #1027 is the one that makes use of it.

Reply:
I see, in that case it's fine 👍